lib/Conversion/PipelineToHW/PipelineToHW.cpp

```diff
@@ -0,0 +1,107 @@
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+//===----------------------------------------------------------------------===//
+//
+// This is the main Pipeline to HW Conversion Pass Implementation.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Conversion/PipelineToHW.h"
+#include "../PassDetail.h"
+#include "circt/Dialect/Comb/CombOps.h"
+#include "circt/Dialect/HW/HWOps.h"
+#include "circt/Dialect/Pipeline/Pipeline.h"
+#include "circt/Dialect/Seq/SeqOps.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+using namespace mlir;
+using namespace circt;
+using namespace pipeline;
+
+class LowerLatencySensitivePipelinePattern
+    : public OpRewritePattern<PipelineOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(PipelineOp pipeline,
+                                PatternRewriter &rewriter) const override {
+    if (pipeline.isLatencyInsensitive())
+      return pipeline.emitOpError() << "Only latency-sensitive pipelines are "
+                                       "supported at the moment";
+
+    // Simply move the ops from the pipeline to the enclosing hw.module scope,
+    // converting any stage ops to seq registers.
+    Value clk = pipeline.clock();
+    Value rst = pipeline.reset();
+    llvm::SmallVector<Value, 4> retVals;
+    rewriter.setInsertionPoint(pipeline);
+
+    for (auto [arg, barg] :
+         llvm::zip(pipeline.getOperands(), pipeline.getBody()->getArguments()))
+      barg.replaceAllUsesWith(arg);
+
+    for (auto &op : llvm::make_early_inc_range(*pipeline.getBody())) {
+      auto loc = op.getLoc();
+      llvm::TypeSwitch<Operation *, void>(&op)
+          .Case<PipelineStageRegisterOp>([&](auto stage) {
+            unsigned stageIdx = stage.index();
+            auto validRegName = rewriter.getStringAttr(
+                "s" + std::to_string(stageIdx) + "_valid");
+            auto validReg = rewriter.create<seq::CompRegOp>(
+                loc, rewriter.getI1Type(), stage.when(), clk, validRegName, rst,
+                Value(), StringAttr());
+            stage.valid().replaceAllUsesWith(validReg);
+
+            for (auto &it : llvm::enumerate(stage.regIns())) {
+              auto regIdx = it.index();
+              auto regIn = it.value();
+              auto regName =
+                  rewriter.getStringAttr("s" + std::to_string(stageIdx) +
+                                         "_reg" + std::to_string(regIdx));
+              auto reg = rewriter.create<seq::CompRegOp>(
+                  loc, regIn.getType(), regIn, clk, regName, rst, Value(),
+                  StringAttr());
+              stage.regOuts()[regIdx].replaceAllUsesWith(reg);
```

Comment:
This and anywhere else you are directly mutating IR in this pattern need to use the PatternRewriter: https://mlir.llvm.org/docs/PatternRewriter/#restrictions.

Reply:
I was obviously grogging this due to the rewriter stuff of `HandshakeToHW` but as far as i can tell, there is no rewriter support for:
- Replacing single SSA values (only `replaceOpWith`-like functions
- Moving single operations (online mergeBlocks, inlineRegion, ...)

Would the right thing here then be to mark all of this stuff as updated-in-place or (more likely) consider this style of transformation as unsuitable for the pattern rewriter.

Reply:
I haven't looked at the details but those are your options: use the methods on PatternRewriter (there's usually something there, but you might have to re-create IR instead of mutating it), use updateRootInPlace, or don't use PatternRewriter.

Reply:
hmm i see... I still sense that this style of implementing the pass is the right way to go (iterate over all ops sequentially to be aware of op ordering) so implementing op-specific rewrite patterns would seem counterintuitive. Given that, and the fact that this leaves us with a single top-level rewrite pattern, gut feel tells me that the right thing to do here is to not use the rewriter infra.